### ISSUE TITLE
Use persistent file watcher to reduce file system reads for unchanged files

### ISSUE THREAD

**Issue by vjpr** [pnpm/pnpm#3346]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

# **Describe the user story**

As a dev,

- I want the cli add and install commands to run as fast as possible.
- If nothing has changed, the command should not do anything.

# **Describe the solution you'd like**

Reading of package manifests is not optimized at present. It involves reading the all workspace manifests from disk.

For 260 workspace packages, it takes around 800ms for me. This runs at both the start and end of the install process (if there are modifications - but maybe this is a bug.)

1. We should only make file system reads for packages that have changed since last operation.

If no manifest has changed, we exit instantly.

2. We should cache the manifests.

## File watcher options

We need a daemon that runs a file-watcher.

### `watchman` 

Maintained by Facebook, works with very large code bases. Doesn't follow symlinks though - but this is not a problem.

PRO: Allows a very simple implementation because we just need to modify `find-packages`. There is a param called `since` that we can pass in to get the latest changes since last run to determine whether to re-read from disk.

### `chokidar` in `pnpm server`

We could add our own file-system watcher in the `pnpm server` daemon.

CON: We have to implement our own watcher logic.

PRO: We have more control over the watcher.

PRO: No external dependency.

## Cache options

We are caching the output of `@pnpm/find-workspace-packages#findPackages`

- In-memory in `pnpm server`
- Single file on disk
  - Would be quite a large file creating a bottleneck...
- No cache
  - We only use the watcher to avoid running phases that don't need to run if no files were changed.

# **Describe the drawbacks of your solution**

- Sometimes `watchman` crashes. We would need to ensure code to avoid getting in to a bad state.

- User has to use `watchman`. We could make it an optional flag initially.

# Interesting

Are there other places we watch to reduce reads? 

- Maybe we run a watcher on the pnpm store to allow instant store mutation checks?

- Can we speed up checks for native deps and other phases?

# Related

https://discord.com/channels/731599538665553971/731599538665553975/831221571519512657

# Benchmark

I created a script that can be run to time the phases.

TODO: Publish the package.

These are results for monorepo with 260 workspace packages.

```
$ pnpm i --reporter=ndjson --loglevel=debug | pnpm-timings

scope 33ms
package-manifest 819ms <---------------- 1
context 33ms
lifecycle 460ms
stats 1ms
stage 78ms
install-check 119ms
skipped-optional-dependency 35ms
install-check 473ms <-------------------- 2
stats 1ms
stage 168ms
link 189ms
global 1ms
link 2.6s <------------------------------ 3
package-manifest 923ms <----------------- 4
```

**Comment by vjpr** [comment]
Looking at implementation options:

- `@pnpm/find-workspace-packages`
  - reads patterns from `pnpm-workspace.yaml`.
  - calls `find-packages`.

- `find-packages`
  - non-pnpm-specific
  - calls `fast-glob` to do disk access.

Wondering what is the best place to add this.

# Current thoughts

`@pnpm/find-workspace-packages` takes `opts.finderImpl` that will be an alternate version of `find-packages`.

**Comment by zkochan** [comment]
Could we use Git to check if the package.json files were changed?

**Comment by vjpr** [comment]
> Could we use Git to check if the package.json files were changed?

Good idea. Sounds familiar...I think we talked about using git to check something similar a few years back. Maybe the store integrity?

**Comment by zkochan** [comment]
Yes, we considered it for the store integrity but that appeared to be slow. But this is a different scenario. Might be good. I don't know.

**Comment by vjpr** [comment]
It would not work for untracked `package.json` files - before their first commit. You would wear the performance hit if you had an untracked `package.json` file. This would be not so common though.

Still has to call `stat` on all your files. Seems to be efficiently implemented with multiple threads. 

https://github.com/git/git/blob/master/builtin/commit.c#L1461

**Comment by zkochan** [comment]
I would not consider `pnpm server`

**Comment by vjpr** [comment]
> I would not consider pnpm server

Why?

**Comment by zkochan** [comment]
pnpm is currently slower with `pnpm server` than without it.

**Comment by vjpr** [comment]
> pnpm is currently slower with pnpm server than without it.

Interesting. Where was the speed-up suppose to come from with `server`?

---

So with a file-based cache it would work as follows when `@pnpm/find-workspace-packages` is called:

1. Read `node_modules/.pnpm/manifests.json`. `{updated: <timestamp>, packages: []}`
2. Call `watchman since <manifests.json#updated>`.
3. Apply updates in-memory.
4. Write to `manifests.json`.
5. Return packages.

**Comment by zkochan** [comment]
Seems like an overkill to constantly run watchman just to get a little boost to the installation speed.

Maybe just reading the stats of files will be enough. And no need for a manifest.json file, there is already a file for meta info: `node_modules/.pnpm/modules.yaml`

> Where was the speed-up suppose to come from with server?

`pnpm server` was added by a member of the Glitch team as they needed a separate process to handle the store. In their env `pnpm install` has not acess to the store. My assymption was that it will also boost installation but my measurements are showing that installation with `pnpm server` is slower than installation without `pnpm server`.

**Comment by vjpr** [comment]
> Seems like an overkill to constantly run watchman just to get a little boost to the installation speed.

Watchman is pretty common. If you use jest it is [enabled for file crawling by default](https://jestjs.io/docs/cli#--watchman). If you use react-native its a necessity.

In my monorepo I have a bunch of things using it.

What's great about it is that it only attaches one watcher per directory, so you can have multiple processes/tools using it. E.g. I use it for generating barrel files, and graphql-codegen when package.json files are modified.

I think it will be good to have as an option to allow the fastest possible installation speed.

**Comment by vjpr** [comment]
Although the more I think about it...maybe there is an inefficiency somewhere that is easier to fix.

I'll do some benchmarks on the globbing, maybe that is causing it. 

Do we always need to read manifests from disk if they haven't changed? 

Could we just run glob, then stat on the pjsons, and read ones that are out of date.

Are we tracking modified times in the .modules.yaml?

Maybe this is the simpler fix.

**Comment by zkochan** [comment]
This is implemented in current pnpm, so closing.

pnpm keeps the workspace scan in node_modules/.pnpm-workspace-state-v1.json and takes an mtime-based fast path before the install pipeline runs (pnpm/crates/package-manager/src/optimistic_repeat_install.rs), which is what this asked a watcher to achieve.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
